{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/BI-BE-CS-183-2023/blob/main/HW8/Problem3.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MozW3XxY8ezr"
      },
      "source": [
        "Bi/Be/Cs 183 2022-2023: Intro to Computational Biology\n",
        "TAs: Meichen Fang, Tara Chari, Zitong (Jerry) Wang\n",
        "\n",
        "**Submit your notebooks by sharing a clickable link with Viewer access. Link must be accessible from submitted assignment document.**\n",
        "\n",
        "Make sure Runtime $\\rightarrow$ Restart and run all works without error"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **HW 8 Problem 3**\n",
        "\n",
        "Chimpanzees are the closest living relatives to humans. Here we will be looking at differences between the two genomes. \n",
        "\n",
        "In this problem you will implement the Needleman-Wunsch algorithm for global alignment between a pair of sequences of human and chimpanzee FOXP2 genes [1]. You will be reading in a FASTA file of DNA sequences and run the algorithm to find their optimal alignment given a set of rewards and penalties for the possible nucleotide pairs. This algorithm allows for gaps in the alignment, and constructs the global optimal alignment by iteratively finding the solutions for sets of small subsequences and their alignments.\n",
        "\n",
        "FOXP2 is a gene that has been implicated in speech and language development in humans, where mutations in the gene can inhibit speech. Interestingly, FOXP2s sequence and function are different between chimpanzees and humans, and its mechanism(s) is an area of study for researchers in understanding how we process and develop language.\n",
        "\n",
        "[1] Enard W, Przeworski M, Fisher SE, Lai CS, Wiebe V, Kitano T, Monaco AP, Pääbo S. Molecular evolution of FOXP2, a gene involved in speech and language. Nature. 2002 Aug 22;418(6900):869-72. doi: 10.1038/nature01025. Epub 2002 Aug 14. PMID: 12192408.\n",
        "\n"
      ],
      "metadata": {
        "id": "Mu4obGyqS8fh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##**Import data and install packages**"
      ],
      "metadata": {
        "id": "fJ8J92ntz2rk"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rjy8W5hly24q"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import scipy.io as sio\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt #Can use other plotting packages like seaborn"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yH-EpSmB9Y_x",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "37d69f70-3790-4928-87ba-3041d75b4bcb"
      },
      "source": [
        "#Download the FASTA file with a pair of nucleotide sequences\n",
        "\n",
        "!wget --content-disposition 'https://drive.google.com/uc?export=download&id=1GFYciWTl2Objrplb1XqNa0pWzTBvXbsm'"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-02-23 09:13:06--  https://drive.google.com/uc?export=download&id=1GFYciWTl2Objrplb1XqNa0pWzTBvXbsm\n",
            "Resolving drive.google.com (drive.google.com)... 172.253.114.113, 172.253.114.139, 172.253.114.101, ...\n",
            "Connecting to drive.google.com (drive.google.com)|172.253.114.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 303 See Other\n",
            "Location: https://doc-0k-3c-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/gmr6j7464c3nav93cmi5scee1qq5iinn/1677143550000/14515314329444727017/*/1GFYciWTl2Objrplb1XqNa0pWzTBvXbsm?e=download&uuid=af2d7ad4-7451-43bf-821f-15400b1906ff [following]\n",
            "Warning: wildcards not supported in HTTP.\n",
            "--2023-02-23 09:13:07--  https://doc-0k-3c-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/gmr6j7464c3nav93cmi5scee1qq5iinn/1677143550000/14515314329444727017/*/1GFYciWTl2Objrplb1XqNa0pWzTBvXbsm?e=download&uuid=af2d7ad4-7451-43bf-821f-15400b1906ff\n",
            "Resolving doc-0k-3c-docs.googleusercontent.com (doc-0k-3c-docs.googleusercontent.com)... 173.194.192.132, 2607:f8b0:4001:c0e::84\n",
            "Connecting to doc-0k-3c-docs.googleusercontent.com (doc-0k-3c-docs.googleusercontent.com)|173.194.192.132|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 49 [application/octet-stream]\n",
            "Saving to: ‘seqs.fasta’\n",
            "\n",
            "seqs.fasta          100%[===================>]      49  --.-KB/s    in 0s      \n",
            "\n",
            "2023-02-23 09:13:07 (3.03 MB/s) - ‘seqs.fasta’ saved [49/49]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Read in data for analysis**"
      ],
      "metadata": {
        "id": "PpeuucqgiDPx"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**The data**"
      ],
      "metadata": {
        "id": "swY2nSDukmXg"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here you will read in  a FASTA file of a pair of DNA sequences. FASTA is a common text file format in which nucleotide sequences are stored. Commonly, the name of the sequence follows a '>'. The following line then has the sequence itself, terminated by an '*'."
      ],
      "metadata": {
        "id": "h9dCxq5twLUk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Print the FASTA file we downloaded\n",
        "!cat seqs.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "njnLSJehwurt",
        "outputId": "73485b91-4c51-4c23-a595-9b788b6a9a0e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">sequence1\n",
            "ACAACAACAG*\n",
            ">sequence2\n",
            "ACAACAGCAGCAG*\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Problem 3 (40 points)**"
      ],
      "metadata": {
        "id": "BSNL7Tr7g2z_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "In the Needleman-Wunsch (NW) algorithm we will have two matrices: the score matrix and traceback matrix. \n",
        "\n",
        "###Score Matrix\n",
        "\n",
        "We begin by constructing a **score matrix** $\\mathbf{F}$, an $(m+1)\\times(n+1)$ matrix for two sequences $x,y$ of length $m$ and $n$ respectively. Each entry $i,j$ corresponds to the alignment score of the nucleotides in at $x_i \\text{ and } y_j$.\n",
        "\n",
        "Nucleotide pairs can either be aligned to eachother (e.g. $F(1,1)$ = 'G' mapped to 'G') or to a gap (e.g. $F(1,0)$ = 'G' mapped to '-'). Pairs can also be aligned if they do not match (e.g. $F(1,2)$ = 'G' mapped to 'A'). (See below for sequences $x$ = 'GATTACA', $y$ = 'GCATGCT')\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "4SveJ1yfd5Ay"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<center><img src=\"https://drive.google.com/uc?export=view&id=1Qfzs96bRkU2YJwx8Ss9VcjRN2EPUyLY4\" alt=\"exGrid\" width=\"300\" height=\"300\"><center>"
      ],
      "metadata": {
        "id": "w9RtLXVrq07n"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "To calculate a score/entry for $F_{i,j}$, we will use the scores of $F_{i-1,j-1},F_{i-1,j}$ and $F_{i,j-1}$. The score for $F_{i,j}$ is defined as the maximum score of (1) $x_i$ being aligned to $y_j$, (2) $x_i$ being aligned to a gap '-', and (3) $y_j$ being aligned to a gap '-'. Thus the score is calculated as:\n",
        "\n",
        "\\begin{align}\n",
        "F_{i,j} = max\\begin{cases} \n",
        "      F_{i-1,j-1} + s(x_i,y_j) & (1) \\\\\n",
        "      F_{i-1,j} - d & (2) \\\\\n",
        "      F_{i,j-1} - d & (3)\n",
        "   \\end{cases}\n",
        "\\end{align}\n",
        "\n",
        "$s()$ denotes a score function, which we will define below, that returns a reward for two matched nucleotides or a penalty for mismatched bases.\n",
        "We will also set a value for the 'gap penalty' $d$."
      ],
      "metadata": {
        "id": "_rfy3lYVq2V5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "We can draw this score calculation as:"
      ],
      "metadata": {
        "id": "A7R6sTansoYU"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<center><img src=\"https://drive.google.com/uc?export=view&id=1MZf_hZ_7Tprj0hV0rpwsBWXBmrmFvWt_\" alt=\"exScore\" width=\"300\" height=\"160\"><center>"
      ],
      "metadata": {
        "id": "Uk_g9dIl1tHT"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "###Traceback Matrix\n",
        "\n",
        "After we find out the highest score, the previous cell which gave the highest candidate score must also be recorded. This information is recorded in the **traceback matrix** $\\mathbf{P}$. For example, if $F_{i-1,j-1} + s(x_i,y_j)$ gives the maximum score for $F_{i,j}$, then the corresponding cell of the traceback matrix $\\mathbf{P}_{ij}$ is ”diag”, which means the last cell is $F_{i-1,j-1}$. If it is the above cell $F_{i-1,j}$ that gives the maximum score , then the $\\mathbf{P}_{ij}$ is ”up”. If there are even scores, we can just record one of them for the problem."
      ],
      "metadata": {
        "id": "TXZSWn_JJOyM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "After filling all cells of the score matrix and traceback matrix, the score in the cell on the bottom right represents the alignment score for the best alignment. We then use the traceback matrix to construct the best alignment.\n",
        "\n",
        "The traceback begins with the bottom right cell. One moves according to the traceback value written in the cell. There are three possible moves: diagonally (toward the top-left corner of the matrix), up, or left.\n",
        "The traceback is completed when the first, top-left cell of\n",
        "the matrix is reached.\n",
        "\n",
        "Here is the bottom right corner of an example traceback matrix:\n",
        "<center><img src=\"https://drive.google.com/uc?export=view&id=183LAe4zPt2s8qsyfVmyScoWxyUBVC6uX\" alt=\"exScore\" width=\"250\" height=\"160\"><center>"
      ],
      "metadata": {
        "id": "wmesEKALUnfS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "In summary, the Needleman-Wunsch algorithm consists of three steps:\n",
        "1. Initialisation of the score matrix\n",
        "2. Calculation of scores and filling the traceback matrix\n",
        "3. Deducing the alignment from the traceback matrix\n",
        "\n",
        "We will implement each step below."
      ],
      "metadata": {
        "id": "74oBMwZuRZP-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **a) Read in the DNA sequences from the FASTA file. Extract just the sequence strings, *not including* the asterisk at the end, and print the two strings. (2 points)**"
      ],
      "metadata": {
        "id": "49ekjwPBijRu"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "OS5HYKVjmeMq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **b) Initialisation of the score matrix: fill in the boundary conditions of $\\mathbf{F}$. (5 points)**\n",
        "The first column and row denote all bases in that column or row being mapped to gaps '-'. Assuming $F_{0,0} = 0$, fill in row 0 and column 0 accordingly, using a **gap penalty $\\mathbf{d = -1}$**. \n",
        "\n",
        "**Print your $\\mathbf{F}$ matrix after initialization.**"
      ],
      "metadata": {
        "id": "hMGOe1xHhP77"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "gj37defNaAYP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **c) Implement the s() (score) function. (3 points)**\n",
        "\n",
        "Fill in the s() function below to take in two nucleotide bases and output a **reward of 2** if the bases match, and a **penalty of -2** for mismatched bases."
      ],
      "metadata": {
        "id": "7rbg1mgJhTYx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "  "
      ],
      "metadata": {
        "id": "ZSlLlmSnhS4A"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **d) Calculation of scores and filling the traceback matrix: fill in the rest of the $\\mathbf{F}$ matrix, row by row, and store the information in traceback matrix $\\mathbf{P}$. (10 points)**\n",
        "\n",
        "Fill in the $\\mathbf{F}$ matrix using the score function described in the main Problem statement, using the s() function from **c)** and a **gap penalty $\\mathbf{d = -1}$**\n",
        " \n",
        "Remember to store the arrow or the previous cells when we fill in the F. \n",
        "\n",
        "**Print your final $\\mathbf{F}$ matrix and $\\mathbf{P}$ matrix.**"
      ],
      "metadata": {
        "id": "I8VzNwjHlFFf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        " "
      ],
      "metadata": {
        "id": "5v07XqO34haj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **e) Beginning at bottom corner, traceback the alignment and output the final alignment of the two sequences (10 points)**"
      ],
      "metadata": {
        "id": "eEz9p0zOlgSu"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "At each step in the traceback we look at the entries of $\\mathbf{P}_{i,j}$ matrix which records which cell (among $F_{i-1,j-1}, F_{i-1,j} and F_{i,j-1}$) $F_{i,j}$ was derived from. We take a 'step' back to the recorded previous entry that gave the max score, and add a pair of symbols/nucleotides to the alignment accordingly.\n",
        "\n",
        "If $\\mathbf{P}_{i,j}$ is diag, the step is to $F_{i-1,j-1}$ (diagonal arrow), $x_i$ and $y_j$ are added to the alignment. \n",
        "\n",
        "If $\\mathbf{P}_{i,j}$ is up, the step is to $F_{i-1,j}$ (vertical arrow), and $x_i$ and '-' (gap character) are added.\n",
        "\n",
        "If $\\mathbf{P}_{i,j}$ is left, the step is to $F_{i,j-1}$ (horizontal arrow), and  '-' and $y_j$ are added."
      ],
      "metadata": {
        "id": "jqRSmq1Ry4Xf"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "If there is a tie for the max entry, choose any of the entries in that tie with equal probability. "
      ],
      "metadata": {
        "id": "GFnVPk3Nl_dC"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Example: here we combine the traceback and score matrix for visualization such that the numbers are the scores and the arrows represent the information in the traceback matrix.\n",
        "\n",
        "Starting at the right corner, according to our recorded arrow, this comes from the left cell, thus T from y and a gap '-' are aligned. In the next step, again based on the arrow, the top cell is selected, thus the A from x is aligned to a gap '-'."
      ],
      "metadata": {
        "id": "h8SXsKENV_8X"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<center><img src=\"https://drive.google.com/uc?export=view&id=1mONEIOXBIBhTlYjevItK6dexR6V091cw\" alt=\"exScore\" width=\"450\" height=\"350\"><center>"
      ],
      "metadata": {
        "id": "SvI2J0E6vV_U"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "The result alignment is:\n",
        "```\n",
        "G\t-\tA\tT\tT\tA\t-\tC\tA\t-\n",
        "G\tC\tA\tT\t-\t-\tG\tC\t-\tT\n",
        "```\n",
        "\n"
      ],
      "metadata": {
        "id": "QbIQn3eGc_dr"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "fTg3dkgumraE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **f) Change the score function to adapt the mismatch penalties to account for more or less likely nucleotide substitutions. (5 points)**\n",
        "\n",
        "Transitions within the purine class (A <--> G) or within the pyrimidine class (C <--> T) are more likely and thus have a **penalty of -1**. Transversions, across classes (e.g. A --> T) have a **penalty of -2** as they are less likely.\n",
        "\n",
        "**Define a new s() function to implement these penalty changes.**"
      ],
      "metadata": {
        "id": "JdE9-1Ormnw2"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "dms3XZ4zmtHD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **g) Re-run the NW algorithm with new penalties, report your new final alignment with the penalty changes/new score function in f), and print the full (final) $\\mathbf{F}$ matrix. (5 points)**\n",
        "\n",
        "Use the adapted mismatch penalties. If there are multiple possible alignments, you just need to report one of them."
      ],
      "metadata": {
        "id": "vSdXasDvnaTY"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "u1qC7GUSFwGb"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}